Support custom Liquibase credentials

appiepollo14 opened the pull request:
- Fixes https://github.com/quarkusio/quarkus/issues/31214

appiepollo14 committed:
Allows to use specific datasource credentials for Liquibase

Fixes https://github.com/quarkusio/quarkus/issues/31214

gsmet commented on .mvn/extensions.xml:
```diff
@@ -19,4 +19,9 @@
         <artifactId>quarkus-project-develocity-extension</artifactId>
         <version>1.0.6</version>
     </extension>
+     <extension>
+         <groupId>org.apache.maven.extensions</groupId>
+         <artifactId>maven-build-cache-extension</artifactId>
+         <version>1.0.0</version>
+     </extension>
```
Could you please revert this change?

gsmet reviewed (changes requested):
This looks awesome but see my small gripe inline.

I will let others review this in details though.

gastaldi commented on .idea/runConfigurations/mvnDebug.xml:
```diff
@@ -4,10 +4,10 @@
     <option name="HOST" value="localhost" />
-    <option name="PORT" value="8000" />
+    <option name="PORT" value="5005" />
     <option name="AUTO_RESTART" value="false" />
     <RunnerSettings RunnerId="Debug">
-      <option name="DEBUG_PORT" value="8000" />
+      <option name="DEBUG_PORT" value="5005" />
```
Please revert this change either

gsmet commented:
As for PR title and commit message, I would go with something like:

> Allow to use specific datasource credentials for Liquibase
>
> Fixes #31214

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:
```diff
@@ -31,9 +34,23 @@ public Liquibase createLiquibase() {
-            ;
+            Database database;
+
+            if (config.username.isPresent() && config.password.isPresent()) {
```
The Javadoc for `username` says: 
- `If no specific JDBC URL is configured, falls back to the datasource username if not specified.`
- 
And for `password`: 
- `If no specific JDBC URL is configured, falls back to the datasource password if not specified.`
 
This seems confusing as they are both ignored if one is not present.

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:
```diff
@@ -31,9 +34,23 @@ public Liquibase createLiquibase() {
+                AgroalDataSource agroalDataSource = (AgroalDataSource) dataSource;
+                String jdbcUrl = agroalDataSource.getConfiguration().connectionPoolConfiguration()
+                        .connectionFactoryConfiguration().jdbcUrl();
+                Connection connection = DriverManager.getConnection(jdbcUrl, config.username.get(), config.password.get());
```
Can we use `datasource.getConnection(String username, String password)` here instead of using the Agroal API?

appiepollo14 replied:
No Than the error, in this https://github.com/quarkusio/quarkus/issues/31214#issuecomment-1991117911  discussion occurs.

appiepollo14 commented on integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java:
```diff
@@ -18,6 +18,20 @@ public void testLiquibaseQuarkusFunctionality() {
+    }
+
+    @Test
+    @DisplayName("Checks the user")
```
@gastaldi question. This test will only succeed if the `testLiquibaseUsingDedicatedUsernameAndPassword` is succeeded. So there is a dependency. Is that allowed?

gastaldi replied:
How about merging the checks with the test above?

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:
```diff
@@ -31,9 +34,23 @@ public Liquibase createLiquibase() {
+            Database database;
+
+            if (config.username.isPresent() && config.password.isPresent()) {
+                AgroalDataSource agroalDataSource = (AgroalDataSource) dataSource;
```
Using `Datasource#unwrap` may be a better option than explicit casting
```suggestion
                AgroalDataSource agroalDataSource = dataSource.unwrap(AgroalDataSource.class);
```

appiepollo14 commented:
@gastaldi refactored. Can this be merged?

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java:
```diff
@@ -80,4 +80,16 @@ public class LiquibaseConfig {
 
+    /**
+     * The username that Liquibase uses to connect to the database.
+     * If no username is configured, falls back to the datasource username.
```
```suggestion
     * If no username is configured, falls back to the datasource username and password.
```

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java:
```diff
@@ -80,4 +80,16 @@ public class LiquibaseConfig {
+
+    /**
+     * The password that Liquibase uses to connect to the database.
+     * If no password is configured, falls back to the datasource password.
```
```suggestion
     * If no password is configured, falls back to the datasource password and username.
```

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java:
```diff
@@ -101,6 +101,20 @@ public static final LiquibaseDataSourceRuntimeConfig defaultConfig() {
 
+    /**
+     * The username that Liquibase uses to connect to the database.
+     * If no specific JDBC URL is configured, falls back to the datasource username if not specified.
```
This description is not really accurate, maybe use the same as above?

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java:
```diff
@@ -101,6 +101,20 @@ public static final LiquibaseDataSourceRuntimeConfig defaultConfig() {
+
+    /**
+     * The password that Liquibase uses to connect to the database.
+     * If no specific JDBC URL is configured, falls back to the datasource password if not specified.
```
Same here

gastaldi reviewed (changes requested):
It looks good overall, just added a minor suggestion to make the config description clearer

gastaldi commented on extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java:
```diff
@@ -80,4 +80,16 @@ public class LiquibaseConfig {
+
+    /**
+     * The password that Liquibase uses to connect to the database.
+     * If no username is configured, falls back to the datasource username and password.
```
```suggestion
     * If no password is configured, falls back to the datasource username and password.
```

appiepollo14 replied:
sorry, too hastely.

gastaldi reviewed (approved):
LGTM, good job!

gastaldi commented:
@appiepollo14 the build failures seem related to your change?

appiepollo14 commented:
@gastaldi It does. Will do some fixing.

appiepollo14 commented:
Disabled one IT on Windows as there is no Docker engine available. Added one extra class to the PMT test. Hopefully that will solve the missing dependency. 

@gastaldi is there a way to run the full CI on my fork?

appiepollo14 commented:
@gastaldi @gsmet this implementation is based on the Flyway implementation. To check the behavior of the extension with custom username / password, a second datasource (postgres) is introduced. This uses devservices to create a DB. The test now fail as, in opposition to the Flyway Integration tests, the PMT has no devservice PostGres database. Can be fixed maybe by reverting to a second datasource in H2, hopefully that has also a current_user default value for a table but should look into that. Other solution would be, to get rid of the PMT for the Liquibase Integration test. I prefer the second solution. Please approve / comment.

gsmet reviewed (approved):
I force pushed to rebase and fix the formatting error.

This patch looks very good and is definitely something we want in.

My only concern is that we should avoid adding a dependency to Hibernate ORM in the tests.
We should rather use Agroal directly and do a database query.

Making everything intertwined in Quarkus makes it very hard to do major changes and we already got hit several times by Hibernate being used all around.

Would you have the cycles to address this concern? It could be after we merged this PR as it's already very good.

Thanks!

appiepollo14 commented:
@gsmet I Will take on that challenge. However, there is still a failing Prod mode test. Please decide and reply to the following comment:

> @gastaldi @gsmet this implementation is based on the Flyway implementation. To check the behavior of the extension with custom username / password, a second datasource (postgres) is introduced. This uses devservices to create a DB. The test now fail as, in opposition to the Flyway Integration tests, the PMT has no devservice PostGres database. Can be fixed maybe by reverting to a second datasource in H2, hopefully that has also a current_user default value for a table but should look into that. Other solution would be, to get rid of the PMT for the Liquibase Integration test. I prefer the second solution. Please approve / comment.

gastaldi commented:
What is PMT?

gastaldi commented:
Maybe you can test with a single datasource, perform an INSERT in the test using the current user and read the table data to check if they match?

appiepollo14 commented:
@gastaldi Please advise

gastaldi commented:
Why do you need a second datasource?

appiepollo14 commented:
Altered code, removed Postgres datasource in favor of H2 second database. Test should now work.

appiepollo14 commented:
@gsmet @gastaldi  removed hibernate and replaced it with agroal. It is partially working . I need te call, I can't figure out why, the endpoint for running Liquibase twice. After the second time I can call the endpoint: "created-by" and retrieve the inserted row from the database. After the first run, I get a database error stating the database is empty. 

This issue was introduced when removing hibernate in favor of agroal sql statement. So I believe it has to do something the agroal datasource. Might be settings?

Hope you guys have the solution so this PR can finally be merged!

gastaldi commented on integration-tests/liquibase/pom.xml:
```diff
@@ -1,7 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Revert the changes in this pom.xml

appiepollo14 commented:
@gastaldi @gsmet 
please guide me by providing an answer to the comment above. https://github.com/quarkusio/quarkus/pull/39361#issuecomment-2028832759

appiepollo14 commented:
Would love to see this pr merged. Help needed solving issue described here: https://github.com/quarkusio/quarkus/pull/39361#issuecomment-2028832759

Any ideas are welcome

gsmet commented:
Hey there. Sorry I was busy on other things :).

I'll have a look now.

gsmet committed:
Adjust the new Liquibase tests a bit

Not sure exactly what was going wrong but I had to create two users to
make things work and also make sure everything was done in one go.

appiepollo14 commented:
@gsmet nice fix!!

gsmet commented:
Merged! Thanks for your work and sorry for the delay!
